x/bsonx/bsoncore/array.go

```diff
@@ -0,0 +1,242 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bsoncore
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strconv"
+)
+
+// Array is a raw bytes representation of a BSON array.
+type Array []byte
+
+// NewArrayFromReader reads an array from r. This function will only validate the length is
+// correct and that the array ends with a null byte.
+func NewArrayFromReader(r io.Reader) (Array, error) {
```

Comment:
A lot of the code for these functions is duplicated from the existing `Document` type (e.g. this function is basically the same as `NewDocumentFromReader`). I can understand why you did it this way, but would it be possible to write helper functions that work with `[]byte` for this kind of thing that both `Document` and `Array` use? I'm thinking something like:

```
func NewArrayFromReader(r io.Reader) (Array, error) {
     return newBufferFromReader(r)
}

func newBufferFromReader(r io.Reader) ([]byte, error) {
    // all the code you have in NewArrayFromReader currently, but with "array" renamed to something like "buffer"
}
```

Reply:
Sure—leaves less code duplication. I renamed a few functions and made them work with `[]byte`: `newBufferFromReader()`, `indexErr()`, and `values()`. There is still some duplication in `bsoncore.Array`'s `DebugString()`, `String()` and `Validate()`, but the functions differ slightly from their `Document` counterparts on key lines.

Reply:
Sure—leaves a lot less code duplication. I reworked a few functions to use `[]byte`: `newBufferFromReader()`, `indexErr()`, and `values()`, and I called them from `array.go` and `document.go`. `bsoncore.Array`'s `DebugString()`, `String()` and `Validate()` still copy some logic from `bsoncore.Document`, but there are critical lines that differ.